File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/async/java/UserService.java
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.async.java;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;

// Java async code is commonly written with CompletableFuture,
// ExecutorService, or platform/virtual threads.
public class UserService {

    private final UserApiClient apiClient;
    private final ExecutorService executorService;

    public UserService(UserApiClient apiClient, ExecutorService executorService) {
        this.apiClient = apiClient;
        this.executorService = executorService;
    }

    public CompletableFuture<User> getUserAsync(String userId) {
        // supplyAsync runs the blocking operation on another thread.
        // The caller receives a CompletableFuture instead of a direct User.
        return CompletableFuture.supplyAsync(
                () -> apiClient.fetchUserById(userId),
                executorService
        );
    }

    public CompletableFuture<String> getUserNameAsync(String userId) {
        // thenApply transforms the result after the async operation completes.
        return getUserAsync(userId)
                .thenApply(User::getName);
    }

    public CompletableFuture<UserProfile> getUserProfileAsync(String userId) {
        // thenCompose is used when another async operation depends
        // on the result of the previous one.
        return getUserAsync(userId)
                .thenCompose(user ->
                        CompletableFuture.supplyAsync(
                                () -> new UserProfile(user, "STANDARD"),
                                executorService
                        )
                );
    }
}

File: backend-systems-lab-stage-1/src/main/java/pl/jakubtworek/backend_systems_lab_stage_1/block_a/kotlin/coroutines/async/kotlin/UserService.kt
package pl.jakubtworek.backend_systems_lab_stage_1.block_a.kotlin.coroutines.async.kotlin

// Kotlin async code is commonly written with suspend functions
// and coroutines instead of CompletableFuture chains.
class UserService(
    private val apiClient: UserApiClient
) {

    suspend fun getUser(userId: String): User {
        // This looks like normal sequential code,
        // but it can suspend without blocking the thread.
        return apiClient.fetchUserById(userId)
    }

    suspend fun getUserName(userId: String): String {
        // No thenApply is needed.
        // The result is handled like a regular value.
        val user = getUser(userId)
        return user.name
    }

    suspend fun getUserProfile(userId: String): UserProfile {
        // Sequential-looking async flow.
        val user = getUser(userId)
        return UserProfile(user, "STANDARD")
    }
}